pkg/bloomcompactor/bloomcompactor.go

```diff
@@ -90,22 +125,335 @@ func New(cfg Config,
+		}
+
+		c.storeClients[periodicConfig.From] = storeClient{
+			object:       objectClient,
+			index:        index_storage.NewIndexStorageClient(objectClient, periodicConfig.IndexTables.PathPrefix),
+			chunk:        chunk_client.NewClient(objectClient, nil, schemaCfg),
+			indexShipper: indexShipper,
+		}
+	}
+
+	c.metrics = newMetrics(r)
+	c.metrics.compactionRunInterval.Set(cfg.CompactionInterval.Seconds())
+
+	c.Service = services.NewBasicService(c.starting, c.running, c.stopping)
 
 	return c, nil
 }
 
-func (c *Compactor) starting(_ context.Context) error {
-	return nil
+func (c *Compactor) starting(_ context.Context) (err error) {
+	c.metrics.compactorRunning.Set(1)
+	return err
+}
+
+func (c *Compactor) running(ctx context.Context) error {
+	// Run an initial compaction before starting the interval.
+	if err := c.runCompaction(ctx); err != nil {
+		level.Error(util_log.Logger).Log("msg", "failed to run compaction", "err", err)
+	}
+
+	ticker := time.NewTicker(util.DurationWithJitter(c.cfg.CompactionInterval, 0.05))
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ticker.C:
+			c.metrics.compactionRunsStarted.Inc()
+			if err := c.runCompaction(ctx); err != nil {
+				c.metrics.compactionRunsErred.Inc()
+				level.Error(util_log.Logger).Log("msg", "failed to run compaction", "err", err)
+				continue
+			}
+			c.metrics.compactionRunsCompleted.Inc()
+		case <-ctx.Done():
+			return nil
+		}
+	}
 }
 
 func (c *Compactor) stopping(_ error) error {
+	c.metrics.compactorRunning.Set(0)
 	return nil
 }
 
+func (c *Compactor) runCompaction(ctx context.Context) error {
+	var (
+		tables []string
+		// it possible for two periods to use the same storage bucket and path prefix (different indexType or schema version)
+		// so more than one index storage client may end up listing the same set of buckets
+		// avoid including the same table twice in the compact tables list.
+		seen = make(map[string]struct{})
+	)
+	for _, sc := range c.storeClients {
+		// refresh index list cache since previous compaction would have changed the index files in the object store
+		sc.index.RefreshIndexTableNamesCache(ctx)
+		tbls, err := sc.index.ListTables(ctx)
+		if err != nil {
+			return fmt.Errorf("failed to list tables: %w", err)
+		}
+
+		for _, table := range tbls {
+			if _, ok := seen[table]; ok {
+				continue
+			}
+
+			tables = append(tables, table)
+			seen[table] = struct{}{}
+		}
+	}
+
+	// process most recent tables first
+	sortTablesByRange(tables)
+
+	// apply passed in compaction limits
+	if c.cfg.SkipLatestNTables <= len(tables) {
+		tables = tables[c.cfg.SkipLatestNTables:]
+	}
+	if c.cfg.TablesToCompact > 0 && c.cfg.TablesToCompact < len(tables) {
+		tables = tables[:c.cfg.TablesToCompact]
+	}
+
+	// Reset discovered tenants metric since we will increase it in compactTable.
+	c.metrics.compactionRunDiscoveredTenants.Set(0)
+
+	parallelism := c.cfg.MaxCompactionParallelism
+	if parallelism == 0 {
+		parallelism = len(tables)
+	}
```

Comment:
The parallelism is configurable and defaults to 1 (with the latest commit). Therefore, this will only happen if we set `MaxCompactionParallelism` to 0 (unlimited).